File: android-tree-sitter/src/main/java/com/itsaky/androidide/treesitter/TSTree.java
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.treesitter;

import com.itsaky.androidide.treesitter.annotations.GenerateNativeHeaders;
import com.itsaky.androidide.treesitter.util.TSObjectFactoryProvider;
import dalvik.annotation.optimization.FastNative;
import java.util.Objects;

public class TSTree extends TSNativeObject {

  protected TSTree(long pointer) {
    super(pointer);
  }

  public static TSTree create(long pointer) {
    return TSObjectFactoryProvider.getFactory().createTree(pointer);
  }

  @Override
  protected void closeNativeObj() {
    Native.delete(getNativeObject());
  }

  /**
   * Get the root node of this tree.
   *
   * @return The root node.
   */
  public TSNode getRootNode() {
    checkAccess();
    return Native.rootNode(getNativeObject());
  }

  /**
   * Get the root node of the syntax tree, but with its position shifted forward by the given
   * offset.
   */
  public TSNode getRootNodeWithOffset(int offsetBytes, TSPoint offsetExtent) {
    checkAccess();
    return Native.rootNodeWithOffset(getNativeObject(), offsetBytes, offsetExtent);
  }

  /**
   * Compare an old edited syntax tree to a new syntax tree representing the same document,
   * returning an array of ranges whose syntactic structure has changed.
   * <p>
   * For this to work correctly, the old syntax tree must have been edited such that its ranges
   * match up to the new tree. Generally, you'll want to call this function right after calling one
   * of the <code>parse*</code> functions in {@link TSParser}. You need to pass the old tree that
   * was passed to parse, as well as the new tree that was returned from that function.
   */
  public TSRange[] getChangedRanges(TSTree oldTree) {
    checkAccess();
    oldTree.checkAccess();

    TSRange[] ranges = Native.changedRanges(getNativeObject(), oldTree.getNativeObject());
    if (ranges == null) {
      return new TSRange[0];
    }
    return ranges;
  }

  /**
   * Get the array of included ranges that was used to parse the syntax tree.
   */
  public TSRange[] getIncludedRanges() {
    checkAccess();

    TSRange[] ranges = Native.includedRanges(getNativeObject());
    if (ranges == null) {
      return new TSRange[0];
    }

    return ranges;
  }

  /**
   * Make a shallow copy of this tree. This is very fast.
   *
   * @return The copy of this tree.
   */
  public TSTree copy() {
    checkAccess();
    return TSTree.create(Native.copy(getNativeObject()));
  }

  /**
   * Notify that this tree has been edited.
   *
   * @param edit The edit.
   */
  public void edit(TSInputEdit edit) {
    Objects.requireNonNull(edit, "TSInputEdit cannot be null");
    checkAccess();
    Native.edit(getNativeObject(), edit);
  }

  /**
   * Get the language that was used to parse the syntax tree.
   *
   * @return The tree sitter language.
   */
  public TSLanguage getLanguage() {
    checkAccess();
    final var langPtr = Native.getLanguage(getNativeObject());
    if (langPtr == 0) {
      return null;
    }

    return TSLanguageCache.get(langPtr);
  }

  @GenerateNativeHeaders(fileName = "tree")
  private static class Native {

    @FastNative
    static native void edit(long tree, TSInputEdit inputEdit);

    @FastNative
    static native void delete(long tree);

    @FastNative
    static native long copy(long tree);

    @FastNative
    static native TSNode rootNode(long tree);

    @FastNative
    static native TSNode rootNodeWithOffset(long tree, int offsetBytes, TSPoint offsetExtent);

    @FastNative
    static native TSRange[] changedRanges(long tree, long oldTree);

    @FastNative
    static native long getLanguage(long tree);

    @FastNative
    static native TSRange[] includedRanges(long nativeObject);
  }
}

File: android-tree-sitter/src/main/cpp/ts_tree.cc
/*
 *  This file is part of android-tree-sitter.
 *
 *  android-tree-sitter library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  android-tree-sitter library is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *  along with android-tree-sitter.  If not, see
 * <https://www.gnu.org/licenses/>.
 */

#include "ts_tree.h"

#include "utils/ts_misc.h"
#include "utils/ts_obj_utils.h"
#include "utils/ts_preconditions.h"

static void TSTree_edit(JNIEnv *env, __TS_ATTR_UNUSED jclass self, jlong tree,
                        jobject inputEdit) {
  req_nnp(env, tree);
  req_nnp(env, inputEdit);
  TSInputEdit edit = _unmarshalInputEdit(env, inputEdit);
  ts_tree_edit((TSTree *)tree, &edit);
}

static void TSTree_delete(JNIEnv *env, __TS_ATTR_UNUSED jclass self, jlong tree) {
  req_nnp(env, tree);
  ts_tree_delete((TSTree *)tree);
}

static jlong TSTree_copy(JNIEnv *env, __TS_ATTR_UNUSED jclass self, jlong tree) {
  req_nnp(env, tree);
  return (jlong)ts_tree_copy((TSTree *)tree);
}

static jobject TSTree_rootNode(JNIEnv *env, __TS_ATTR_UNUSED jclass self, jlong tree) {
  req_nnp(env, tree);
  return _marshalNode(env, ts_tree_root_node((TSTree *)tree));
}

static jobject TSTree_rootNodeWithOffset(JNIEnv *env,
                                         __TS_ATTR_UNUSED jclass self,
                                         jlong tree, jint offset_bytes,
                                         jobject offset_extent) {
  req_nnp(env, tree);
  return _marshalNode(
      env, ts_tree_root_node_with_offset((TSTree *)tree, offset_bytes,
                                         _unmarshalPoint(env, offset_extent)));
}

static jobjectArray TSTree_changedRanges(JNIEnv *env,
                                         __TS_ATTR_UNUSED jclass self,
                                         jlong tree, jlong oldTree) {
  req_nnp(env, tree, "thisTree");
  req_nnp(env, oldTree, "oldTree");
  uint32_t count;
  TSRange *ranges =
      ts_tree_get_changed_ranges((TSTree *)oldTree, (TSTree *)tree, &count);
  if (count == 0) {
    if (ranges != nullptr) {
      free(ranges);
    }
    return nullptr;
  }

  jobjectArray arr = createRangeArr(env, (jint)count);
  req_nnp(env, arr, "TSRange[] from factory");

  for (uint32_t i = 0; i < count; i++) {
    TSRange *r = (ranges + i);
    env->SetObjectArrayElement(arr, (jint)i, _marshalRange(env, *r));
  }

  free(ranges);

  return arr;
}

static jobjectArray TSTree_includedRanges(JNIEnv *env,
                                          __TS_ATTR_UNUSED jclass self,
                                          jlong tree) {
  req_nnp(env, tree);

  uint32_t count;
  TSRange *ranges = ts_tree_included_ranges((TSTree *)tree, &count);
  if (count == 0) {
    if (ranges != nullptr) {
      free(ranges);
    }

    return nullptr;
  }

  jobjectArray arr = createRangeArr(env, (jint)count);
  req_nnp(env, arr, "TSRange[] from factory");

  for (uint32_t i = 0; i < count; i++) {
    TSRange *r = (ranges + i);
    env->SetObjectArrayElement(arr, (jint)i, _marshalRange(env, *r));
  }

  free(ranges);

  return arr;
}

static jlong TSTree_getLanguage(JNIEnv *env, __TS_ATTR_UNUSED jclass self,
                                jlong tree) {
  req_nnp(env, tree);
  return (jlong)ts_tree_language((TSTree *)tree);
}

void TSTree_Native__SetJniMethods(JNINativeMethod *methods,
                                  __TS_ATTR_UNUSED int count) {
  SET_JNI_METHOD(methods, TSTree_Native_edit, TSTree_edit)
  SET_JNI_METHOD(methods, TSTree_Native_delete, TSTree_delete)
  SET_JNI_METHOD(methods, TSTree_Native_copy, TSTree_copy)
  SET_JNI_METHOD(methods, TSTree_Native_rootNode, TSTree_rootNode)
  SET_JNI_METHOD(methods, TSTree_Native_rootNodeWithOffset,
                 TSTree_rootNodeWithOffset)
  SET_JNI_METHOD(methods, TSTree_Native_changedRanges, TSTree_changedRanges)
  SET_JNI_METHOD(methods, TSTree_Native_includedRanges, TSTree_includedRanges)
  SET_JNI_METHOD(methods, TSTree_Native_getLanguage, TSTree_getLanguage)
}